File: src/creds.rs
//! Read-only readers for the OAuth credential files the official Claude and
//! Codex CLIs maintain.
//!
//! DESIGN RULE: this app never writes these files and never refreshes tokens.
//! Refreshing would rotate the refresh-token out from under the user's CLI and
//! risk logging them out of their AI tools. We only *read* the access token; if
//! it has already expired we report [`CredState::Expired`] and let the user
//! re-authenticate with their own CLI (`claude` / `codex login`).

use std::path::Path;

use base64::Engine;
use serde::Deserialize;

/// Outcome of reading a credential file.
pub enum CredState<T> {
    /// Valid, non-expired token.
    Valid(T),
    /// Token present but already expired — user must re-login via their CLI.
    Expired,
    /// File missing.
    Missing,
    /// File present but unparseable.
    Malformed(String),
}

// ---------------------------------------------------------------------------
// Anthropic — ~/.claude/.credentials.json
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct AnthropicCreds {
    pub access_token: String,
    pub plan_label: String,
}

#[derive(Deserialize)]
struct AnthropicFile {
    #[serde(rename = "claudeAiOauth")]
    claude_ai_oauth: AnthropicOauth,
}

#[derive(Deserialize)]
struct AnthropicOauth {
    #[serde(rename = "accessToken")]
    access_token: String,
    #[serde(rename = "expiresAt", default, deserialize_with = "de_num")]
    expires_at_ms: i64,
    #[serde(rename = "subscriptionType", default)]
    subscription_type: String,
    #[serde(rename = "rateLimitTier", default)]
    rate_limit_tier: String,
}

/// Read Anthropic credentials. `now_secs` is injected for testability.
pub fn read_anthropic(path: &Path, now_secs: i64) -> CredState<AnthropicCreds> {
    let raw = match std::fs::read_to_string(path) {
        Ok(r) => r,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return CredState::Missing,
        Err(e) => return CredState::Malformed(e.to_string()),
    };
    let file: AnthropicFile = match serde_json::from_str(&raw) {
        Ok(f) => f,
        Err(e) => return CredState::Malformed(e.to_string()),
    };
    let o = file.claude_ai_oauth;
    // Strict expiry: only flag once actually past expiry (we never refresh).
    if o.expires_at_ms > 0 && o.expires_at_ms / 1000 <= now_secs {
        return CredState::Expired;
    }
    CredState::Valid(AnthropicCreds {
        access_token: o.access_token,
        plan_label: anthropic_plan_label(&o.subscription_type, &o.rate_limit_tier),
    })
}

fn anthropic_plan_label(sub: &str, tier: &str) -> String {
    let mut name = capitalize(sub);
    if name.is_empty() {
        name = "Unknown".into();
    }
    if tier.contains("20x") {
        name.push_str(" 20x");
    } else if tier.contains("5x") {
        name.push_str(" 5x");
    }
    name
}

// ---------------------------------------------------------------------------
// OpenAI Codex — ~/.codex/auth.json
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct OpenAiCreds {
    pub access_token: String,
    pub account_id: Option<String>,
    pub plan_hint: Option<String>,
}

#[derive(Deserialize)]
struct OpenAiFile {
    tokens: OpenAiTokens,
}

#[derive(Deserialize)]
struct OpenAiTokens {
    access_token: String,
    id_token: String,
    #[serde(default)]
    account_id: Option<String>,
}

/// Read Codex credentials. Expiry + plan tier come from the id_token JWT.
pub fn read_openai(path: &Path, now_secs: i64) -> CredState<OpenAiCreds> {
    let raw = match std::fs::read_to_string(path) {
        Ok(r) => r,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return CredState::Missing,
        Err(e) => return CredState::Malformed(e.to_string()),
    };
    let file: OpenAiFile = match serde_json::from_str(&raw) {
        Ok(f) => f,
        Err(e) => return CredState::Malformed(e.to_string()),
    };
    let t = file.tokens;
    let claims = parse_jwt_claims(&t.id_token);
    let exp = claims
        .as_ref()
        .and_then(|c| c.get("exp"))
        .and_then(|v| v.as_i64())
        .unwrap_or(0);
    // exp == 0 means unparseable — attempt the fetch anyway; a 401 will then
    // surface as a re-login prompt. Only flag Expired when we KNOW it's past.
    if exp > 0 && exp <= now_secs {
        return CredState::Expired;
    }
    let plan_hint = claims.as_ref().and_then(|c| {
        c.get("https://api.openai.com/auth")
            .and_then(|v| v.get("chatgpt_plan_type"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
    });
    CredState::Valid(OpenAiCreds {
        access_token: t.access_token,
        account_id: t.account_id,
        plan_hint,
    })
}

fn parse_jwt_claims(token: &str) -> Option<serde_json::Value> {
    let payload = token.split('.').nth(1)?;
    let decoded = base64::engine::general_purpose::URL_SAFE_NO_PAD
        .decode(payload)
        .or_else(|_| base64::engine::general_purpose::URL_SAFE.decode(payload))
        .ok()?;
    serde_json::from_slice(&decoded).ok()
}

// ---------------------------------------------------------------------------

fn de_num<'de, D>(d: D) -> Result<i64, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let v = serde_json::Value::deserialize(d)?;
    Ok(match v {
        serde_json::Value::Number(n) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)).unwrap_or(0),
        _ => 0,
    })
}

fn capitalize(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        Some(c) => c.to_uppercase().collect::<String>() + chars.as_str(),
        None => String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::NamedTempFile;

    fn tmp(s: &str) -> NamedTempFile {
        let mut f = NamedTempFile::new().unwrap();
        f.write_all(s.as_bytes()).unwrap();
        f.flush().unwrap();
        f
    }

    fn jwt(claims: serde_json::Value) -> String {
        let h = base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(b"{}");
        let p =
            base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(claims.to_string().as_bytes());
        format!("{h}.{p}.sig")
    }

    #[test]
    fn anthropic_valid_token() {
        let f = tmp(
            r#"{"claudeAiOauth":{"accessToken":"AT","refreshToken":"RT",
                "expiresAt":2000000000000,"subscriptionType":"max",
                "rateLimitTier":"default_claude_max_20x"}}"#,
        );
        match read_anthropic(f.path(), 1_000_000_000) {
            CredState::Valid(c) => {
                assert_eq!(c.access_token, "AT");
                assert_eq!(c.plan_label, "Max 20x");
            }
            _ => panic!("expected valid"),
        }
    }

    #[test]
    fn anthropic_expired_is_flagged_not_refreshed() {
        // expiresAt in the past → Expired, never a network refresh.
        let f = tmp(
            r#"{"claudeAiOauth":{"accessToken":"AT","refreshToken":"RT",
                "expiresAt":1000,"subscriptionType":"pro","rateLimitTier":""}}"#,
        );
        assert!(matches!(
            read_anthropic(f.path(), 1_000_000_000),
            CredState::Expired
        ));
    }

    #[test]
    fn anthropic_missing_file() {
        let p = std::path::Path::new("/nonexistent/aub-win/.credentials.json");
        assert!(matches!(read_anthropic(p, 0), CredState::Missing));
    }

    #[test]
    fn openai_valid_with_plan_hint() {
        let id = jwt(serde_json::json!({
            "exp": 2_000_000_000,
            "https://api.openai.com/auth": {"chatgpt_plan_type": "plus"}
        }));
        let body = format!(
            r#"{{"tokens":{{"access_token":"AT","refresh_token":"RT","id_token":"{id}","account_id":"acc"}}}}"#
        );
        let f = tmp(&body);
        match read_openai(f.path(), 1_000_000_000) {
            CredState::Valid(c) => {
                assert_eq!(c.access_token, "AT");
                assert_eq!(c.account_id.as_deref(), Some("acc"));
                assert_eq!(c.plan_hint.as_deref(), Some("plus"));
            }
            _ => panic!("expected valid"),
        }
    }

    #[test]
    fn openai_expired() {
        let id = jwt(serde_json::json!({"exp": 1000}));
        let body =
            format!(r#"{{"tokens":{{"access_token":"AT","refresh_token":"RT","id_token":"{id}"}}}}"#);
        let f = tmp(&body);
        assert!(matches!(
            read_openai(f.path(), 1_000_000_000),
            CredState::Expired
        ));
    }

    #[test]
    fn openai_unparseable_exp_attempts_anyway() {
        let body =
            r#"{"tokens":{"access_token":"AT","refresh_token":"RT","id_token":"not.a.jwt"}}"#;
        let f = tmp(body);
        // exp unknown → Valid (we try; a live 401 would prompt re-login).
        assert!(matches!(
            read_openai(f.path(), 1_000_000_000),
            CredState::Valid(_)
        ));
    }
}

File: AiUsageBar/Services/Creds.cs
using System;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace AiUsageBar.Services;

/// <summary>
/// Readers for the OAuth credential files the official Claude and Codex CLIs
/// maintain.
///
/// DEFAULT: strictly read-only. We read the access token; if it has already
/// expired we report <see cref="CredKind.Expired"/> and let the user
/// re-authenticate with their own CLI. This never touches the files.
///
/// OPT-IN REFRESH: when the user enables it (<see cref="Config.RefreshEnabled"/>),
/// the <c>...Async</c> readers will, on a near-expiry token, call the OAuth
/// endpoint and write the rotated tokens back to the same credential file the
/// CLI reads — sharing one source of truth instead of forking a stale copy.
/// This rotates the refresh token, so a CLI mid-session may need to re-login;
/// that trade-off is surfaced in the settings UI before the user turns it on.
/// </summary>
public enum CredKind { Valid, Expired, Missing, Malformed }

public sealed class CredResult<T>
{
    public CredKind Kind { get; }
    public T? Value { get; }
    public string? Error { get; }

    private CredResult(CredKind kind, T? value, string? error)
    {
        Kind = kind;
        Value = value;
        Error = error;
    }

    public static CredResult<T> Valid(T value) => new(CredKind.Valid, value, null);
    public static CredResult<T> Expired() => new(CredKind.Expired, default, null);
    public static CredResult<T> Missing() => new(CredKind.Missing, default, null);
    public static CredResult<T> Malformed(string error) => new(CredKind.Malformed, default, error);
}

public sealed record AnthropicCreds(string AccessToken, string PlanLabel);

public sealed record OpenAiCreds(string AccessToken, string? AccountId, string? PlanHint);

public static class Creds
{
    // -- Anthropic — ~/.claude/.credentials.json ------------------------------

    /// <summary>Raw on-disk OAuth fields, before any expiry/refresh policy.</summary>
    private sealed record AnthropicRaw(
        string AccessToken, string RefreshToken, long ExpiresAtMs, string Sub, string Tier);

    private static AnthropicRaw? ReadAnthropicRaw(string path, out string? error)
    {
        error = null;
        string raw;
        try { raw = File.ReadAllText(path); }
        catch (Exception e) { error = e.Message; return null; }

        try
        {
            using var doc = JsonDocument.Parse(raw);
            if (!doc.RootElement.TryGetProperty("claudeAiOauth", out var o))
            {
                error = "missing claudeAiOauth";
                return null;
            }
            return new AnthropicRaw(
                GetString(o, "accessToken") ?? "",
                GetString(o, "refreshToken") ?? "",
                GetLong(o, "expiresAt"),
                GetString(o, "subscriptionType") ?? "",
                GetString(o, "rateLimitTier") ?? "");
        }
        catch (Exception e)
        {
            error = e.Message;
            return null;
        }
    }

    /// <summary>Read-only: never refreshes. Flags <see cref="CredKind.Expired"/>
    /// once the token is past expiry.</summary>
    public static CredResult<AnthropicCreds> ReadAnthropic(string path, long nowSecs)
    {
        if (!File.Exists(path)) return CredResult<AnthropicCreds>.Missing();

        var c = ReadAnthropicRaw(path, out var error);
        if (c is null) return CredResult<AnthropicCreds>.Malformed(error ?? "unknown error");

        // Strict expiry: only flag once actually past expiry (we never refresh).
        if (c.ExpiresAtMs > 0 && c.ExpiresAtMs / 1000 <= nowSecs)
            return CredResult<AnthropicCreds>.Expired();

        return CredResult<AnthropicCreds>.Valid(
            new AnthropicCreds(c.AccessToken, AnthropicPlanLabel(c.Sub, c.Tier)));
    }

    /// <summary>Refresh-aware reader. Falls back to the read-only path when
    /// <paramref name="refreshEnabled"/> is false. When enabled and the token is
    /// near expiry, it refreshes via the OAuth endpoint and writes the rotated
    /// tokens back to disk (best-effort) before returning the fresh access
    /// token.</summary>
    public static async Task<CredResult<AnthropicCreds>> ReadAnthropicAsync(
        string path, long nowSecs, bool refreshEnabled, HttpClient http)
    {
        if (!refreshEnabled) return ReadAnthropic(path, nowSecs);
        if (!File.Exists(path)) return CredResult<AnthropicCreds>.Missing();

        var c = ReadAnthropicRaw(path, out var error);
        if (c is null) return CredResult<AnthropicCreds>.Malformed(error ?? "unknown error");

        var plan = AnthropicPlanLabel(c.Sub, c.Tier);
        var expiresSecs = c.ExpiresAtMs > 0 ? c.ExpiresAtMs / 1000 : 0;

        // Fresh enough, or no refresh token to use: keep the existing token.
        if (!OAuthClient.NeedsRefresh(expiresSecs, nowSecs) || string.IsNullOrEmpty(c.RefreshToken))
            return ResolveAnthropic(c.AccessToken, plan, expiresSecs, nowSecs);

        var r = await OAuthClient.RefreshAnthropicAsync(http, c.RefreshToken).ConfigureAwait(false);
        if (r.Ok && r.AccessToken is { } at)
        {
            var newRefresh = string.IsNullOrEmpty(r.RefreshToken) ? c.RefreshToken : r.RefreshToken!;
            var newExpiresMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
                               + (r.ExpiresInSecs ?? 3600) * 1000;
            try { WriteAnthropicBack(path, at, newRefresh, newExpiresMs); }
            catch { /* best-effort: the refresh worked, so still serve fresh data */ }
            return CredResult<AnthropicCreds>.Valid(new AnthropicCreds(at, plan));
        }

        // Refresh failed: the existing token may still be valid for a few minutes.
        return ResolveAnthropic(c.AccessToken, plan, expiresSecs, nowSecs);
    }

    /// <summary>Decide what to return for a non-refreshed Anthropic token: still
    /// usable, or already expired.</summary>
    private static CredResult<AnthropicCreds> ResolveAnthropic(
        string accessToken, string plan, long expiresSecs, long nowSecs)
        => expiresSecs > 0 && expiresSecs <= nowSecs
            ? CredResult<AnthropicCreds>.Expired()
            : CredResult<AnthropicCreds>.Valid(new AnthropicCreds(accessToken, plan));

    /// <summary>Merge rotated tokens into the existing credentials JSON, preserving
    /// every other field the Claude CLI keeps there. Atomic.</summary>
    private static void WriteAnthropicBack(string path, string access, string refresh, long expiresAtMs)
    {
        var root = LoadJsonObject(path);
        if (root["claudeAiOauth"] is not JsonObject oauth)
        {
            oauth = new JsonObject();
            root["claudeAiOauth"] = oauth;
        }
        oauth["accessToken"] = access;
        oauth["refreshToken"] = refresh;
        oauth["expiresAt"] = expiresAtMs;
        WriteJsonAtomic(path, root);
    }

    private static string AnthropicPlanLabel(string sub, string tier)
    {
        var name = Capitalize(sub);
        if (string.IsNullOrEmpty(name)) name = "Unknown";
        if (tier.Contains("20x")) name += " 20x";
        else if (tier.Contains("5x")) name += " 5x";
        return name;
    }

    // -- OpenAI Codex — ~/.codex/auth.json ------------------------------------

    /// <summary>Raw on-disk OpenAI Codex token fields. <c>ExpSecs</c> is parsed
    /// from the id_token JWT (0 when unparseable).</summary>
    private sealed record OpenAiRaw(
        string AccessToken, string RefreshToken, string IdToken,
        string? AccountId, long ExpSecs, string? PlanHint);

    private static OpenAiRaw? ReadOpenAiRaw(string path, out string? error)
    {
        error = null;
        string raw;
        try { raw = File.ReadAllText(path); }
        catch (Exception e) { error = e.Message; return null; }

        try
        {
            using var doc = JsonDocument.Parse(raw);
            if (!doc.RootElement.TryGetProperty("tokens", out var t))
            {
                error = "missing tokens";
                return null;
            }

            var idToken = GetString(t, "id_token") ?? "";
            var (exp, planHint) = ParseIdToken(idToken);
            return new OpenAiRaw(
                GetString(t, "access_token") ?? "",
                GetString(t, "refresh_token") ?? "",
                idToken,
                GetString(t, "account_id"),
                exp,
                planHint);
        }
        catch (Exception e)
        {
            error = e.Message;
            return null;
        }
    }

    /// <summary>Extract <c>exp</c> (Unix seconds, 0 if absent) and the ChatGPT
    /// plan-type hint from an id_token JWT.</summary>
    private static (long Exp, string? PlanHint) ParseIdToken(string idToken)
    {
        var claims = ParseJwtClaims(idToken);
        if (claims is not { } c) return (0, null);
        var exp = GetLong(c, "exp");
        string? planHint = null;
        if (c.TryGetProperty("https://api.openai.com/auth", out var auth)
            && auth.ValueKind == JsonValueKind.Object)
        {
            planHint = GetString(auth, "chatgpt_plan_type");
        }
        return (exp, planHint);
    }

    /// <summary>Read-only: never refreshes.</summary>
    public static CredResult<OpenAiCreds> ReadOpenAi(string path, long nowSecs)
    {
        if (!File.Exists(path)) return CredResult<OpenAiCreds>.Missing();

        var c = ReadOpenAiRaw(path, out var error);
        if (c is null) return CredResult<OpenAiCreds>.Malformed(error ?? "unknown error");

        // exp == 0 means unparseable — attempt the fetch anyway; a 401 will then
        // surface as a re-login prompt. Only flag Expired when past.
        if (c.ExpSecs > 0 && c.ExpSecs <= nowSecs) return CredResult<OpenAiCreds>.Expired();

        return CredResult<OpenAiCreds>.Valid(new OpenAiCreds(c.AccessToken, c.AccountId, c.PlanHint));
    }

    /// <summary>Refresh-aware reader; see <see cref="ReadAnthropicAsync"/>.</summary>
    public static async Task<CredResult<OpenAiCreds>> ReadOpenAiAsync(
        string path, long nowSecs, bool refreshEnabled, HttpClient http)
    {
        if (!refreshEnabled) return ReadOpenAi(path, nowSecs);
        if (!File.Exists(path)) return CredResult<OpenAiCreds>.Missing();

        var c = ReadOpenAiRaw(path, out var error);
        if (c is null) return CredResult<OpenAiCreds>.Malformed(error ?? "unknown error");

        // Refresh when near expiry, or when the expiry is unknown (exp == 0).
        var needsRefresh = c.ExpSecs <= 0 || OAuthClient.NeedsRefresh(c.ExpSecs, nowSecs);
        if (!needsRefresh || string.IsNullOrEmpty(c.RefreshToken))
            return ResolveOpenAi(c.AccessToken, c.AccountId, c.PlanHint, c.ExpSecs, nowSecs);

        var r = await OAuthClient.RefreshOpenAiAsync(http, c.RefreshToken).ConfigureAwait(false);
        if (r.Ok && r.AccessToken is { } at)
        {
            var newRefresh = string.IsNullOrEmpty(r.RefreshToken) ? c.RefreshToken : r.RefreshToken!;
            var newId = string.IsNullOrEmpty(r.IdToken) ? c.IdToken : r.IdToken!;
            try { WriteOpenAiBack(path, at, newRefresh, newId); }
            catch { /* best-effort */ }
            // Re-derive the plan hint from the new id_token when we got one.
            var planHint = ParseIdToken(newId).PlanHint ?? c.PlanHint;
            return CredResult<OpenAiCreds>.Valid(new OpenAiCreds(at, c.AccountId, planHint));
        }

        return ResolveOpenAi(c.AccessToken, c.AccountId, c.PlanHint, c.ExpSecs, nowSecs);
    }

    private static CredResult<OpenAiCreds> ResolveOpenAi(
        string accessToken, string? accountId, string? planHint, long expSecs, long nowSecs)
        => expSecs > 0 && expSecs <= nowSecs
            ? CredResult<OpenAiCreds>.Expired()
            : CredResult<OpenAiCreds>.Valid(new OpenAiCreds(accessToken, accountId, planHint));

    /// <summary>Merge rotated tokens into <c>auth.json</c>, preserving every other
    /// field the Codex CLI keeps there. Atomic.</summary>
    private static void WriteOpenAiBack(string path, string access, string refresh, string idToken)
    {
        var root = LoadJsonObject(path);
        if (root["tokens"] is not JsonObject tokens)
        {
            tokens = new JsonObject();
            root["tokens"] = tokens;
        }
        tokens["access_token"] = access;
        tokens["refresh_token"] = refresh;
        tokens["id_token"] = idToken;
        // The Codex CLI stamps this on its own refreshes; mirror it.
        root["last_refresh"] = DateTimeOffset.UtcNow.ToString("o");
        WriteJsonAtomic(path, root);
    }

    private static JsonElement? ParseJwtClaims(string token)
    {
        var parts = token.Split('.');
        if (parts.Length < 2) return null;
        try
        {
            var bytes = DecodeBase64Url(parts[1]);
            // Returned JsonElement must outlive the JsonDocument, so clone it.
            using var doc = JsonDocument.Parse(bytes);
            return doc.RootElement.Clone();
        }
        catch
        {
            return null;
        }
    }

    private static byte[] DecodeBase64Url(string s)
    {
        var b = s.Replace('-', '+').Replace('_', '/');
        switch (b.Length % 4)
        {
            case 2: b += "=="; break;
            case 3: b += "="; break;
        }
        return Convert.FromBase64String(b);
    }

    // -- lenient JSON helpers (mirror the tolerant Rust deserializers) --------

    private static string? GetString(JsonElement obj, string name)
        => obj.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String
            ? v.GetString()
            : null;

    private static long GetLong(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v) || v.ValueKind != JsonValueKind.Number) return 0;
        if (v.TryGetInt64(out var i)) return i;
        return v.TryGetDouble(out var d) ? (long)d : 0;
    }

    private static string Capitalize(string s)
    {
        if (string.IsNullOrEmpty(s)) return string.Empty;
        return char.ToUpperInvariant(s[0]) + s[1..];
    }

    // -- write-back helpers (only reached on opt-in refresh) ------------------

    /// <summary>Parse an existing JSON file into a mutable object, tolerating a
    /// missing/garbage file by starting fresh — so a write-back never throws on
    /// read and always lands a valid document.</summary>
    private static JsonObject LoadJsonObject(string path)
    {
        try
        {
            if (File.Exists(path) && JsonNode.Parse(File.ReadAllText(path)) is JsonObject obj)
                return obj;
        }
        catch { /* fall through to a fresh object */ }
        return new JsonObject();
    }

    /// <summary>Write <paramref name="root"/> via a temp file + atomic replace, so
    /// a crash mid-write can't truncate the credential file the CLI depends on.</summary>
    private static void WriteJsonAtomic(string path, JsonObject root)
    {
        var json = root.ToJsonString(new JsonSerializerOptions { WriteIndented = true });
        var tmp = path + ".tmp";
        File.WriteAllText(tmp, json);
        File.Move(tmp, path, overwrite: true);
    }
}
